File: Baekjoon/Class4/03.15654 N과 M (5).cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int n, m;
vector<char> visited;
vector<int> track;
vector<int> nums;

void back_track() {
    if (track.size() == m) { // 출력
        for (int num : track)
            cout << num << ' ';
        cout << '\n';
        return;
    }
    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            track.push_back(nums[i]); // 추가
            visited[i] = true;
            back_track();
            track.pop_back(); // 삭제
            visited[i] = false;
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> n >> m;
    visited = vector<char>(n);
    nums = vector<int>(n);
    for (int i = 0; i < n; i++)
        cin >> nums[i];
    sort(nums.begin(), nums.end()); // 정렬
    back_track();
    return 0;
}

File: Baekjoon/Class4/03.15654 N과 M (5).java
import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {
    private static int n, m;
    private static boolean[] visited;
    private static int[] track;
    private static int[] nums;
    private static StringBuilder sb = new StringBuilder();

    private static void backTrack(int depth) {
        if(depth == m) { // 출력
            for(int num : track)
                sb.append(num).append(' ');
            sb.append('\n');
            return;
        }
        for(int i = 0; i < n; i++) {
            if(!visited[i]) {
                track[depth] = nums[i]; // 추가
                visited[i] = true;
                backTrack(depth + 1);
                visited[i] = false; // 삭제
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        visited = new boolean[n];
        track = new int[m];
        nums = new int[n];
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < n; i++)
            nums[i] = Integer.parseInt(st.nextToken());
        Arrays.sort(nums); // 정렬

        backTrack(0);
        System.out.print(sb);
        br.close();
    }
}